bin/ch-completion.bash

```diff
@@ -0,0 +1,403 @@
+                    --password-many -s --storage --tls-no-verify
+                    -v --verbose --version"
+
+_image_subcommands="build build-cache delete gestalt
+                    import list pull push reset undelete"
+
+# archs taken from ARCH_MAP in charliecloud.py
+_archs="amd64 arm/v5 arm/v6 arm/v7 arm64/v8 386 mips64le ppc64le s390x"
+
+## ch-image ##
+
+# Completion function for ch-image
+#
+_ch_image_complete () {
+    local prev
+    local cur
+    local words
+    local sub_cmd
+    local strg_dir
+    local extras=
+    _get_comp_words_by_ref -n : cur prev words
+
+    # To find the subcommand and storage directory, we pass the associated
+    # functions the current command line minus the last word (note that
+    # “${words[@]::${#words[@]}-1}” in bash is analagous to “words[:-1]” in
+    # python). We do this because the last word is most likely either an empty
+    # string, or is not yet complete. We don't lose anything by dropping the
+    # empty string, and an incomplete word in the command line can lead to false
+    # positives from these functions and consequently unexpected behavior, so we
+    # don't consider it.
+    sub_cmd=$(_ch_image_subcmd_get "${words[@]::${#words[@]}-1}")
+    strg_dir=$(_ch_find_storage "${words[@]::${#words[@]}-1}")
+
+    # Populate debug log
+    if [[ -n "$CH_COMPLETION_DEBUG" ]]; then
+        # shellcheck disable=SC2129
+        echo "\$ ${words[*]}" >> /tmp/ch-completion.log
+        echo "  current: $cur" >> /tmp/ch-completion.log
+        echo "  previous: $prev" >> /tmp/ch-completion.log
+        echo "  sub command: $sub_cmd" >> /tmp/ch-completion.log
+    fi
+
+    # Common opts that take args
+    #
+    case "$prev" in
+    -a | --arch )
+        COMPREPLY=( $(compgen -W "host yolo $_archs" -- "$cur") )
+        return 0
+        ;;
+    --cache-large )
+        # This is just a user-specified number. Can't autocomplete
+        COMPREPLY=()
+        return 0
+        ;;
+    -s | --storage )
+        # This “if” helps avoid overzealous completion. E.g. if there's only one
+        # subdir of the current dir, this command completes to that dir even if
+        # "$cur" is empty (i.e. the user hasn't yet typed anything that could
+        # generate a completion). I didn't like that, hence the “if”.
+        if [[ -n "$cur" ]]; then
+            compopt -o nospace
+            COMPREPLY=( $(compgen -d -S / -- "$cur") )
+        fi
+        return 0
+        ;;
+    esac
+
+    case "$sub_cmd" in
+    build )
+        # FIXME: opts with args that I need to revisit for this logic
+        #
+        #   * --build-arg KEY[=VALUE]
+        #   * --force-cmd=CMD,ARG1[,ARG2...]
+        #
+        case "$prev" in
+        # Go through a list of potential subcommand-specific opts to see if
+        # “$cur” should be an argument. Otherwise, default to CONTEXT or any
+        # valid option (common or subcommand-specific).
+        -f|--file )
```

Comment:
I've not seen the space before the paren before. I suspect it's not idiomatic???

```suggestion
        -f|--file)
```